api/handler/multipart_upload.go

```diff
@@ -0,0 +1,549 @@
+	uploadID := uuid.New()
+
+	additional := []zap.Field{zap.String("uploadID", uploadID.String()), zap.String("Key", reqInfo.ObjectName)}
+	uploadInfo := &layer.UploadInfoParams{
+		UploadID: uploadID.String(),
+		Bkt:      bktInfo,
+		Key:      reqInfo.ObjectName,
+	}
+
+	p := &layer.UploadPartParams{
+		Info:       uploadInfo,
+		PartNumber: 0,
+		Header:     metadata,
+		Reader:     r.Body,
+		Size:       r.ContentLength,
+	}
+
+	info, err := h.obj.UploadPart(r.Context(), p)
+	if err != nil {
+		h.logAndSendError(w, "could not upload a part", reqInfo, err, additional...)
+	}
+
+	var newEaclTable *eacl.Table
+	if containsACLHeaders(r) {
+		if newEaclTable, err = h.getNewEAclTable(r, info); err != nil {
+			h.logAndSendError(w, "could not get new eacl table", reqInfo, err, additional...)
+			return
+		}
+	}
+	if newEaclTable != nil {
+		p := &layer.PutBucketACLParams{
+			Name: reqInfo.BucketName,
+			EACL: newEaclTable,
+		}
+
+		if err = h.obj.PutBucketACL(r.Context(), p); err != nil {
+			h.logAndSendError(w, "could not put bucket acl", reqInfo, err, additional...)
+			if err = h.obj.AbortMultipartUpload(r.Context(), uploadInfo); err != nil {
+				h.log.Error("could not delete multipart upload init file", additional...)
+			}
+			return
+		}
+	}
+
+	tagSet, err := parseTaggingHeader(r.Header)
+	if tagSet != nil {
+		if err = h.obj.PutObjectTagging(r.Context(), &layer.PutTaggingParams{ObjectInfo: info, TagSet: tagSet}); err != nil {
+			h.logAndSendError(w, "could not upload object tagging", reqInfo, err, additional...)
+			if err = h.obj.AbortMultipartUpload(r.Context(), uploadInfo); err != nil {
+				h.log.Error("could not delete multipart upload init file", additional...)
+			}
+			return
+		}
+	}
+
+	resp := InitiateMultipartUploadResponse{
+		Bucket:   bktInfo.Name,
+		Key:      reqInfo.ObjectName,
+		UploadID: uploadID.String(),
+	}
+
+	if err := api.EncodeToResponse(w, resp); err != nil {
+		h.logAndSendError(w, "could not encode InitiateMultipartUploadResponse to response", reqInfo, err, additional...)
+		return
+	}
+}
+
+func (h *handler) UploadPartHandler(w http.ResponseWriter, r *http.Request) {
+	reqInfo := api.GetReqInfo(r.Context())
+
+	bktInfo, err := h.obj.GetBucketInfo(r.Context(), reqInfo.BucketName)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket info", reqInfo, err)
+		return
+	}
+
+	if err := checkOwner(bktInfo, r.Header.Get(api.AmzExpectedBucketOwner)); err != nil {
+		h.logAndSendError(w, "expected owner doesn't match", reqInfo, err)
+		return
+	}
+
+	queryValues := r.URL.Query()
+
+	partNum, err := strconv.Atoi(queryValues.Get(partNumberHeaderName))
+	if err != nil || partNum == 0 {
+		h.logAndSendError(w, "invalid part number", reqInfo, err)
+		return
+	}
+
+	p := &layer.UploadPartParams{
+		Info: &layer.UploadInfoParams{
+			UploadID: queryValues.Get(uploadIdHeaderName),
+			Bkt:      bktInfo,
+			Key:      reqInfo.ObjectName,
+		},
+		PartNumber: partNum,
+		Size:       r.ContentLength,
```

Comment:
We should check part size according to [limits](https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html)